build/build.go

```diff
@@ -298,45 +319,176 @@ func augmentOriginalImports(importPath string, file *ast.File) {
-					d.Name = ast.NewIdent("_")
+					file.Decls[i] = nil
+				}
+			} else if recvKey := astutil.FuncReceiverKey(d); len(recvKey) > 0 {
+				// check if the receiver has been purged, if so, remove the method too.
+				if info, ok := overrides[recvKey]; ok && info.purgeMethods {
+					file.Decls[i] = nil
 				}
 			}
 		case *ast.GenDecl:
-			for _, spec := range d.Specs {
+			for j, spec := range d.Specs {
 				switch s := spec.(type) {
 				case *ast.TypeSpec:
 					if _, ok := overrides[s.Name.Name]; ok {
-						s.Name = ast.NewIdent("_")
-						// Change to struct type with no type body and not type parameters.
-						s.Type = &ast.StructType{Struct: s.Pos(), Fields: &ast.FieldList{}}
-						s.TypeParams = nil
+						d.Specs[j] = nil
 					}
 				case *ast.ValueSpec:
-					for i, name := range s.Names {
-						if _, ok := overrides[name.Name]; ok {
-							s.Names[i] = ast.NewIdent("_")
+					if len(s.Names) == len(s.Values) {
+						// multi-value context
+						// e.g. var a, b = 2, foo[int]()
+						// A removal will also remove the value which may be from a
+						// function call. This allows us to remove unwanted statements.
+						// However, if that call has a side effect which still needs
+						// to be run, add the call into the overlay.
+						for k, name := range s.Names {
+							if _, ok := overrides[name.Name]; ok {
+								s.Names[k] = nil
+								s.Values[k] = nil
+							}
+						}
+					} else {
+						// single-value context
+						// e.g. var a, b = foo[int]()
+						// If a removal from the overlays makes all returned values unused,
+						// then remove the function call as well. This allows us to stop
+						// unwanted calls if needed. If that call has a side effect which
+						// still needs to be run, add the call into the overlay.
+						nameRemoved := false
+						for _, name := range s.Names {
+							if _, ok := overrides[name.Name]; ok {
+								nameRemoved = true
+								name.Name = `_`
+							}
+						}
+						if nameRemoved {
+							removeSpec := true
+							for _, name := range s.Names {
+								if name.Name != `_` {
+									removeSpec = false
+									break
+								}
+							}
+							if removeSpec {
+								d.Specs[j] = nil
+							}
 						}
 					}
 				}
 			}
 		}
 	}
+	finalizeRemovals(file)
+}
+
+// pruneImports will remove any unused imports from the file.
+//
+// This will not remove any dot (`.`) or blank (`_`) imports.
+// If the removal of code causes an import to be removed, the init's from that
+// import may not be run anymore. If we still need to run an init for an import
+// which is no longer used, add it to the overlay as a blank (`_`) import.
+func pruneImports(file *ast.File) {
+	unused := make(map[string]int, len(file.Imports))
+	for i, in := range file.Imports {
+		if name := astutil.ImportName(in); len(name) > 0 {
+			unused[name] = i
+		}
+	}
+
+	// Remove "unused import" for any import which is used.
+	ast.Walk(astutil.NewCallbackVisitor(func(n ast.Node) bool {
```

Comment:
Isn't this the same as https://pkg.go.dev/go/ast#Inspect?

Reply:
Yep, you're correct. I'll change over too `Inspect`.

I based this code off of [golang.org/x/tools/imports](https://pkg.go.dev/golang.org/x/tools/imports) and it used `ast.Walk` (seen [here](https://cs.opensource.google/go/x/tools/+/refs/tags/v0.16.1:internal/imports/fix.go;l=188)). So that was in my head when I wrote this lightweight version. `Inspect` would definitely be better here. Thanks for catching that.

Reply:
Fixed